core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -442,6 +446,22 @@ public void setMessageCompression(boolean enabled) {
     stream.setMessageCompression(enabled);
   }
 
+  @Override
+  public Cork cork() {
```

Comment:
I don't think we should have any part of this implementation be thread-safe. By necessity the caller has to coordinate with senders and sending is not thread-safe.

Reply:
I'm not sure about this one. I'm pretty sure many users would close cork after certain timeout, which would be called from different thread, so they would have to do synchronization on their own, but that may be ok since whole ClientCallImpl is not thread-safe.

Reply:
Agreed with the batch processing, but this is also useful for stream processing, where user does not now if another write is coming, but still wants to trade of throughput for latency.

Anyway, I can remove the atomic properties and user can provide own synchronization if necessary.

What do you think?

Reply:
Yeah, let's do that. The `stream.flush();` here isn't actually thread-safe and the complexity involved to change it would be high and limit other optimizations.